src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/StringSearchValuesHelper.cs

```diff
@@ -0,0 +1,294 @@
+                return false;
+            }
+
+            return TCaseSensitivity.Equals(ref matchStart, candidate);
+        }
+
+        public interface IValueLength
+        {
+            static abstract bool AtLeast4Chars { get; }
+            static abstract bool AtLeast8Chars { get; }
+        }
+
+        public readonly struct ValueLengthLessThan4 : IValueLength
+        {
+            public static bool AtLeast4Chars => false;
+            public static bool AtLeast8Chars => false;
+        }
+
+        public readonly struct ValueLength4To7 : IValueLength
+        {
+            public static bool AtLeast4Chars => true;
+            public static bool AtLeast8Chars => false;
+        }
+
+        public readonly struct ValueLength8OrLonger : IValueLength
+        {
+            public static bool AtLeast4Chars => true;
+            public static bool AtLeast8Chars => true;
+        }
+
+        public interface ICaseSensitivity
+        {
+            static abstract char TransformInput(char input);
+            static abstract Vector128<byte> TransformInput(Vector128<byte> input);
+            static abstract Vector256<byte> TransformInput(Vector256<byte> input);
+            static abstract Vector512<byte> TransformInput(Vector512<byte> input);
+            static abstract bool Equals(ref char matchStart, string candidate);
+            static abstract bool Equals<TValueLength>(ref char matchStart, string candidate) where TValueLength : struct, IValueLength;
+        }
+
+        public readonly struct CaseSensitive : ICaseSensitivity
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static char TransformInput(char input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector128<byte> TransformInput(Vector128<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector256<byte> TransformInput(Vector256<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static Vector512<byte> TransformInput(Vector512<byte> input) => input;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool Equals(ref char matchStart, string candidate)
+            {
+                ref char end = ref Unsafe.Add(ref matchStart, candidate.Length);
+                ref char candidateRef = ref Unsafe.AsRef(candidate.GetPinnableReference());
+
+                do
+                {
+                    if (candidateRef != matchStart)
+                    {
+                        return false;
+                    }
+
+                    matchStart = ref Unsafe.Add(ref matchStart, 1);
+                    candidateRef = ref Unsafe.Add(ref candidateRef, 1);
+                }
+                while (Unsafe.IsAddressLessThan(ref matchStart, ref end));
+
+                return true;
+            }
```

Comment:
Is this much better than:
```C#
for (int i = 0; i < candidate.Length; i++)
    if (candidate[i] != Unsafe.Add(ref matchStart, i))
        return false;
return true;
```
?

Reply:
No, doesn't seem to be.
These were all changed to use a shared helper now to avoid some duplication.